torch/csrc/autograd/graph_task.h

```diff
@@ -31,6 +31,8 @@ struct GraphTask : std::enable_shared_from_this<GraphTask> {
   std::unordered_map<Node*, InputBuffer> not_ready_;
   std::unordered_map<Node*, int> dependencies_;
 
+  // Records the nodes that are in the graph
+  std::unordered_set<Node*> nodes_in_graph_;
```

Comment:
To access `dependencies_` we'd need to acquire the graph task mutex. That might be problematic because we need to acquire the same mutex when writing to graph_task->captured_vars_ and call user-defined capture hooks. If the user-defined code in those capture hooks calls into our new API that would lead to a deadlock.

Reply:
Reading dependencies_ requires the mutex to be acquired, and someone might try to call this API when the mutex is already acquired, e.g. when capture variable hooks are called.

Reply:
We never add new elements to the map though, we only change some values. Is it still unsafe to check if a key is in the map in that context?